### ISSUE TITLE
Dead connection never die

### ISSUE THREAD

**Issue by gabriel-tessier** [elastic/elasticsearch-py#562]
Hi,

I use plugin 5.2.0 and making some test, if you take a fresh install and set 2 servers let's say 127.0.0.1:9200 (who work) and 1.2.3.4:9200 (which not exist), I always have a timeout exception and my site have a latency (about the timeout set), somehow the 1.2.3.4 is not marked a dead and continue to try to connect, I tried with 3 ips (2 are not existing) and the latency is double etc...

Here the stack trace that I have with 60~70% of my requests if I refresh 10 times in ~30 sec.
For this test I set timeout to 1 cause I don't want to wait.
```
WARNING:elasticsearch:GET http://1.2.3.4:9200/_nodes/_all/http [status:N/A request:1.002s]
Traceback (most recent call last):
  File "/home/src/lib/elasticsearch/connection/http_urllib3.py", line 114, in perform_request
    response = self.pool.urlopen(method, url, body, retries=False, headers=self.headers, **kw)
  File "/home/src/lib/urllib3/connectionpool.py", line 640, in urlopen
    _stacktrace=sys.exc_info()[2])
  File "/home/src/lib/urllib3/util/retry.py", line 238, in increment
    raise six.reraise(type(error), error, _stacktrace)
  File "/home/src/lib/urllib3/connectionpool.py", line 595, in urlopen
    chunked=chunked)
  File "/home/src/lib/urllib3/connectionpool.py", line 363, in _make_request
    conn.request(method, url, **httplib_request_kw)
  File "/usr/local/lib/python2.7/httplib.py", line 1053, in request
    self._send_request(method, url, body, headers)
  File "/usr/local/lib/python2.7/httplib.py", line 1093, in _send_request
    self.endheaders(body)
  File "/usr/local/lib/python2.7/httplib.py", line 1049, in endheaders
    self._send_output(message_body)
  File "/usr/local/lib/python2.7/httplib.py", line 893, in _send_output
    self.send(msg)
  File "/usr/local/lib/python2.7/httplib.py", line 855, in send
    self.connect()
  File "/home/src/lib/urllib3/connection.py", line 167, in connect
    conn = self._new_conn()
  File "/home/src/lib/urllib3/connection.py", line 147, in _new_conn
    (self.host, self.timeout))
ConnectTimeoutError: (<urllib3.connection.HTTPConnection object at 0x7fc74dbb4fd0>, u'Connection to 1.2.3.4 timed out. (connect timeout=1)')
```

Here the full command I use with all the params:
```
elasticsearch_server = ('127.0.0.1:9200', '1.2.3.4:9200')
flask.g.es = elasticsearch.Elasticsearch(elasticsearch_server, timeout=1, retry_on_timeout=True, sniff_on_start=True, sniff_on_connection_fail=True, serializer=JSONSerializerPython2(), sniffer_timeout=60, sniff_timeout=10)
```

the timeout=1 is only for this test on production is 10.

If I remove the not working IP from the list (i.e 1.2.3.4:9200)  I have no error and no latency!

If I have 2 or 3 servers and one die I'll always have the latency!!! TT

Thanks for any help!

**Comment by HonzaKral** [comment]
Hi @gabriel-tessier, sorry for the late reply, I have bee trying to replicate your situation or understand what might be happening.

I am unable to replicate your situation, every time I put in a non-existent IP address it gets (correctly) put on a 60s timeout and only the live connection is being used even when I used all the parameters you mention in the ticket.

Could you please try this again, watch the logs for messages like `"Connection <Urllib3HttpConnection: http://1.2.3.4:9200> has failed for 1 times in a row, putting on 60 second timeout."` and inspect the list of living connections after every request (`print(es.transport.connection_pool.connections)`)

Thank you!

**Comment by HonzaKral** [comment]
This is what I see, which is the correct behavior:
```
In [1]: from elasticsearch import Elasticsearch
In [2]: es = Elasticsearch(["127.0.0.1", "1.2.3.4"], retry_on_timeout=True, sniff_on_connection_fail=True, sniffer_timeout=60, timout=1)
In [3]: print(es.transport.connection_pool.connections)
[<Urllib3HttpConnection: http://1.2.3.4:9200>, <Urllib3HttpConnection: http://127.0.0.1:9200>]
In [4]: es.ping()
HEAD http://1.2.3.4:9200/ [status:N/A request:3.146s]
Traceback (most recent call last):
  File "/home/honza/work/VENVS/es3/lib/python3.6/site-packages/urllib3/connection.py", line 138, in _new_conn
    (self.host, self.port), self.timeout, **extra_kw)
  File "/home/honza/work/VENVS/es3/lib/python3.6/site-packages/urllib3/util/connection.py", line 98, in create_connection
    raise err
  File "/home/honza/work/VENVS/es3/lib/python3.6/site-packages/urllib3/util/connection.py", line 88, in create_connection
    sock.connect(sa)
OSError: [Errno 101] Network is unreachable

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/honza/work/elastic/elasticsearch-py/elasticsearch/connection/http_urllib3.py", line 114, in perform_request
    response = self.pool.urlopen(method, url, body, retries=False, headers=self.headers, **kw)
  File "/home/honza/work/VENVS/es3/lib/python3.6/site-packages/urllib3/connectionpool.py", line 643, in urlopen
    _stacktrace=sys.exc_info()[2])
  File "/home/honza/work/VENVS/es3/lib/python3.6/site-packages/urllib3/util/retry.py", line 311, in increment
    raise six.reraise(type(error), error, _stacktrace)
  File "/home/honza/work/VENVS/es3/lib/python3.6/site-packages/urllib3/packages/six.py", line 686, in reraise
    raise value
  File "/home/honza/work/VENVS/es3/lib/python3.6/site-packages/urllib3/connectionpool.py", line 594, in urlopen
    chunked=chunked)
  File "/home/honza/work/VENVS/es3/lib/python3.6/site-packages/urllib3/connectionpool.py", line 361, in _make_request
    conn.request(method, url, **httplib_request_kw)
  File "/usr/lib64/python3.6/http/client.py", line 1239, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/usr/lib64/python3.6/http/client.py", line 1285, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/usr/lib64/python3.6/http/client.py", line 1234, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/usr/lib64/python3.6/http/client.py", line 1026, in _send_output
    self.send(msg)
  File "/usr/lib64/python3.6/http/client.py", line 964, in send
    self.connect()
  File "/home/honza/work/VENVS/es3/lib/python3.6/site-packages/urllib3/connection.py", line 163, in connect
    conn = self._new_conn()
  File "/home/honza/work/VENVS/es3/lib/python3.6/site-packages/urllib3/connection.py", line 147, in _new_conn
    self, "Failed to establish a new connection: %s" % e)
urllib3.exceptions.NewConnectionError: <urllib3.connection.HTTPConnection object at 0x7f73540d5940>: Failed to establish a new connection: [Errno 101] Network is unreachable
Connection <Urllib3HttpConnection: http://1.2.3.4:9200> has failed for 1 times in a row, putting on 60 second timeout.
Out[4]: True
In [5]: print(es.transport.connection_pool.connections)
(<Urllib3HttpConnection: http://127.0.0.1:9200>,)
In [6]: es.ping()
Out[6]: True
In [7]: es.ping()
Out[7]: True
In [8]: es.ping()
Out[8]: True
```

**Comment by gabriel-tessier** [comment]
Thanks for the reply, I'm checking now.

**Comment by HonzaKral** [comment]
btw I see that you are using `sniff_on_start` which would mean that the connection list (`.transport.connection_pool.connections`) will actually get replaced with a new list obtained from the cluster itself so the original wrong IP should be completely removed and used only when sniffing while all other connections are dead.

let me know if there is anything I can do; and apologies again for the late reply

**Comment by gabriel-tessier** [comment]
I tried again, actually I only change the timeout to 60 but still have the same latency I refresh 3 times and get 2 times timeout so somehow it continue to try to connect to the not existing IP.
Yes the list of connection is always correct but the app still try to connect to the non existing IP.

Here the connection commande:
```
elasticsearch_server = ('127.0.0.1:9200', '1.2.3.4:9200')
es = elasticsearch.Elasticsearch(elasticsearch_server, timeout=60, retry_on_timeout=True, sniff_on_start=True, sniff_on_connection_fail=True, serializer=JSONSerializerPython2(), sniffer_timeout=60, sniff_timeout=10)
```
Here my logs:
```
[10:09:52.670364] DONE
WARNING:elasticsearch:GET http://1.2.3.4:9200/_nodes/_all/http [status:N/A request:60.065s]
Traceback (most recent call last):
  File "/home/src/lib/elasticsearch/connection/http_urllib3.py", line 114, in perform_request
    response = self.pool.urlopen(method, url, body, retries=False, headers=self.headers, **kw)
  File "/home/src/lib/urllib3/connectionpool.py", line 640, in urlopen
    _stacktrace=sys.exc_info()[2])
  File "/home/src/lib/urllib3/util/retry.py", line 238, in increment
    raise six.reraise(type(error), error, _stacktrace)
  File "/home/src/lib/urllib3/connectionpool.py", line 595, in urlopen
    chunked=chunked)
  File "/home/src/lib/urllib3/connectionpool.py", line 363, in _make_request
    conn.request(method, url, **httplib_request_kw)
  File "/usr/local/lib/python2.7/httplib.py", line 1053, in request
    self._send_request(method, url, body, headers)
  File "/usr/local/lib/python2.7/httplib.py", line 1093, in _send_request
    self.endheaders(body)
  File "/usr/local/lib/python2.7/httplib.py", line 1049, in endheaders
    self._send_output(message_body)
  File "/usr/local/lib/python2.7/httplib.py", line 893, in _send_output
    self.send(msg)
  File "/usr/local/lib/python2.7/httplib.py", line 855, in send
    self.connect()
  File "/home/src/lib/urllib3/connection.py", line 167, in connect
    conn = self._new_conn()
  File "/home/src/lib/urllib3/connection.py", line 147, in _new_conn
    (self.host, self.timeout))
ConnectTimeoutError: (<urllib3.connection.HTTPConnection object at 0x7fd468e11650>, u'Connection to 1.2.3.4 timed out. (connect timeout=60)')
(<Urllib3HttpConnection: http://127.0.0.1:9200>,)
WARNING:elasticsearch:GET http://1.2.3.4:9200/_nodes/_all/http [status:N/A request:60.052s]
Traceback (most recent call last):
  File "/home/src/lib/elasticsearch/connection/http_urllib3.py", line 114, in perform_request
    response = self.pool.urlopen(method, url, body, retries=False, headers=self.headers, **kw)
  File "/home/src/lib/urllib3/connectionpool.py", line 640, in urlopen
    _stacktrace=sys.exc_info()[2])
  File "/home/src/lib/urllib3/util/retry.py", line 238, in increment
    raise six.reraise(type(error), error, _stacktrace)
  File "/home/src/lib/urllib3/connectionpool.py", line 595, in urlopen
    chunked=chunked)
  File "/home/src/lib/urllib3/connectionpool.py", line 363, in _make_request
    conn.request(method, url, **httplib_request_kw)
  File "/usr/local/lib/python2.7/httplib.py", line 1053, in request
    self._send_request(method, url, body, headers)
  File "/usr/local/lib/python2.7/httplib.py", line 1093, in _send_request
    self.endheaders(body)
  File "/usr/local/lib/python2.7/httplib.py", line 1049, in endheaders
    self._send_output(message_body)
  File "/usr/local/lib/python2.7/httplib.py", line 893, in _send_output
    self.send(msg)
  File "/usr/local/lib/python2.7/httplib.py", line 855, in send
    self.connect()
  File "/home/src/lib/urllib3/connection.py", line 167, in connect
    conn = self._new_conn()
  File "/home/src/lib/urllib3/connection.py", line 147, in _new_conn
    (self.host, self.timeout))
ConnectTimeoutError: (<urllib3.connection.HTTPConnection object at 0x7fd468e1d6d0>, u'Connection to 1.2.3.4 timed out. (connect timeout=60)')
(<Urllib3HttpConnection: http://127.0.0.1:9200>,)
127.0.0.1 - - [13/Apr/2017 10:10:29] "GET /favicon.ico HTTP/1.0" 404 -
INFO:werkzeug:127.0.0.1 - - [13/Apr/2017 10:10:29] "GET /favicon.ico HTTP/1.0" 404 -
127.0.0.1 - - [13/Apr/2017 10:10:29] "GET /favicon.ico HTTP/1.0" 404 -
INFO:werkzeug:127.0.0.1 - - [13/Apr/2017 10:10:29] "GET /favicon.ico HTTP/1.0" 404 -
Exception in thread Thread-8:
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/threading.py", line 810, in __bootstrap_inner
    self.run()
  File "/usr/local/lib/python2.7/threading.py", line 763, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/usr/local/lib/python2.7/SocketServer.py", line 602, in process_request_thread
    self.handle_error(request, client_address)
  File "/usr/local/lib/python2.7/SocketServer.py", line 599, in process_request_thread
    self.finish_request(request, client_address)
  File "/usr/local/lib/python2.7/SocketServer.py", line 334, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/usr/local/lib/python2.7/SocketServer.py", line 655, in __init__
    self.handle()
  File "/home/src/lib/werkzeug/serving.py", line 216, in handle
    rv = BaseHTTPRequestHandler.handle(self)
  File "/usr/local/lib/python2.7/BaseHTTPServer.py", line 340, in handle
    self.handle_one_request()
  File "/home/src/lib/werkzeug/serving.py", line 251, in handle_one_request
    return self.run_wsgi()
  File "/home/src/lib/werkzeug/serving.py", line 193, in run_wsgi
    execute(self.server.app)
  File "/home/src/lib/werkzeug/serving.py", line 184, in execute
    write(data)
  File "/home/src/lib/werkzeug/serving.py", line 152, in write
    self.send_header(key, value)
  File "/usr/local/lib/python2.7/BaseHTTPServer.py", line 401, in send_header
    self.wfile.write("%s: %s\r\n" % (keyword, value))
IOError: [Errno 32] Broken pipe

Exception in thread Thread-9:
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/threading.py", line 810, in __bootstrap_inner
    self.run()
  File "/usr/local/lib/python2.7/threading.py", line 763, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/usr/local/lib/python2.7/SocketServer.py", line 602, in process_request_thread
    self.handle_error(request, client_address)
  File "/usr/local/lib/python2.7/SocketServer.py", line 599, in process_request_thread
    self.finish_request(request, client_address)
  File "/usr/local/lib/python2.7/SocketServer.py", line 334, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/usr/local/lib/python2.7/SocketServer.py", line 655, in __init__
    self.handle()
  File "/home/src/lib/werkzeug/serving.py", line 216, in handle
    rv = BaseHTTPRequestHandler.handle(self)
  File "/usr/local/lib/python2.7/BaseHTTPServer.py", line 340, in handle
    self.handle_one_request()
  File "/home/src/lib/werkzeug/serving.py", line 251, in handle_one_request
    return self.run_wsgi()
  File "/home/src/lib/werkzeug/serving.py", line 193, in run_wsgi
    execute(self.server.app)
  File "/home/src/lib/werkzeug/serving.py", line 184, in execute
    write(data)
  File "/home/src/lib/werkzeug/serving.py", line 152, in write
    self.send_header(key, value)
  File "/usr/local/lib/python2.7/BaseHTTPServer.py", line 401, in send_header
    self.wfile.write("%s: %s\r\n" % (keyword, value))
IOError: [Errno 32] Broken pipe

[10:10:52.725271] EXECUTE: u'SELECT `user`.created_at, `user`.deleted_at, `user`.extra, `user`.id, `user`.mail, `user`.status, `user`.updated_at FROM `user` WHERE `user`.id = %s LIMIT 1', (240135,)
[10:10:52.725712] DONE
(<Urllib3HttpConnection: http://127.0.0.1:9200>,)
127.0.0.1 - - [13/Apr/2017 10:10:52] "GET /favicon.ico HTTP/1.0" 404 -
WARNING:elasticsearch:GET http://1.2.3.4:9200/_nodes/_all/http [status:N/A request:60.063s]
Traceback (most recent call last):
  File "/home/src/lib/elasticsearch/connection/http_urllib3.py", line 114, in perform_request
    response = self.pool.urlopen(method, url, body, retries=False, headers=self.headers, **kw)
  File "/home/src/lib/urllib3/connectionpool.py", line 640, in urlopen
    _stacktrace=sys.exc_info()[2])
  File "/home/src/lib/urllib3/util/retry.py", line 238, in increment
    raise six.reraise(type(error), error, _stacktrace)
  File "/home/src/lib/urllib3/connectionpool.py", line 595, in urlopen
    chunked=chunked)
  File "/home/src/lib/urllib3/connectionpool.py", line 363, in _make_request
    conn.request(method, url, **httplib_request_kw)
  File "/usr/local/lib/python2.7/httplib.py", line 1053, in request
    self._send_request(method, url, body, headers)
  File "/usr/local/lib/python2.7/httplib.py", line 1093, in _send_request
    self.endheaders(body)
  File "/usr/local/lib/python2.7/httplib.py", line 1049, in endheaders
    self._send_output(message_body)
  File "/usr/local/lib/python2.7/httplib.py", line 893, in _send_output
    self.send(msg)
  File "/usr/local/lib/python2.7/httplib.py", line 855, in send
    self.connect()
  File "/home/src/lib/urllib3/connection.py", line 167, in connect
    conn = self._new_conn()
  File "/home/src/lib/urllib3/connection.py", line 147, in _new_conn
    (self.host, self.timeout))
ConnectTimeoutError: (<urllib3.connection.HTTPConnection object at 0x7fd4683ab1d0>, u'Connection to 1.2.3.4 timed out. (connect timeout=60)')
INFO:werkzeug:127.0.0.1 - - [13/Apr/2017 10:10:52] "GET /favicon.ico HTTP/1.0" 404 -
(<Urllib3HttpConnection: http://127.0.0.1:9200>,)
(<Urllib3HttpConnection: http://127.0.0.1:9200>,)
(<Urllib3HttpConnection: http://127.0.0.1:9200>,)
(<Urllib3HttpConnection: http://127.0.0.1:9200>,)
(<Urllib3HttpConnection: http://127.0.0.1:9200>,)
(<Urllib3HttpConnection: http://127.0.0.1:9200>,)
(<Urllib3HttpConnection: http://127.0.0.1:9200>,)
(<Urllib3HttpConnection: http://127.0.0.1:9200>,)
(<Urllib3HttpConnection: http://127.0.0.1:9200>,)
(<Urllib3HttpConnection: http://127.0.0.1:9200>,)
(<Urllib3HttpConnection: http://127.0.0.1:9200>,)
```

**Comment by gabriel-tessier** [comment]
Side question maybe not related but the urllib version is '1.16' is there some incompatibility or elasticsearch-py need a specific version?
also python is 2.7
I can reproduce the problem on my local and on some dev server on AWS.
Maybe the problem can come from other lib.

**Comment by HonzaKral** [comment]
Where and how are you creating the connection? Can you recreate this in a single script outside of your server?

I see that you are using threads which makes me immediately suspicious whether you are actually using the same `Elasticsearch` object in each instance. Also it looks like everything stabilized for you in the end where you only see the `127.0.0.1` connection in the end so you shouldn't see any errors...

**Comment by gabriel-tessier** [comment]
I checked in unit test and it work as expected, creating only one connection then using the correct connection to make all the call.
But I use it for a web site and each time somebody make a call it create a new connection (+ other new connection from thread for some ajax call).
So each time somebody visit the page he can hit the not responding server and have timeout. In my test it's 40~70% of the requests (for 2 servers).

I thought that the lib was persisting the list of active connection somewhere and not only for the active session. Because there's also the sniffer_timeout that I set to 60sec, but if a active session is time to serve a page it's always less than 1 sec and if it reinitialize for each call 60 sec have no mean.
Is it a good idea to save the list of active connection in memcached?

I also update urllib to 1.20 and it change nothing.

Thanks again for the help.
